glib/src/date.rs

```diff
@@ -130,73 +152,125 @@ impl Date {
+        if !Self::valid_dmy(self.get_day(), month, self.get_year()) {
+            Err(bool_error!("invalid month"))
+        } else {
+            unsafe {
+                ffi::g_date_set_month(self.to_glib_none_mut().0, month.to_glib());
+            }
+            Ok(())
         }
     }
 
-    pub fn set_parse(&mut self, str: &str) {
-        unsafe {
-            ffi::g_date_set_parse(self.to_glib_none_mut().0, str.to_glib_none().0);
+    pub fn set_parse(&mut self, str: &str) -> Result<(), BoolError> {
+        let mut c = self.clone();
+        if !unsafe {
+            ffi::g_date_set_parse(c.to_glib_none_mut().0, str.to_glib_none().0);
+            ffi::g_date_valid(c.to_glib_none().0) == 0
+        } {
+            Err(bool_error!("invalid parse string"))
+        } else {
+            *self = c;
+            Ok(())
         }
     }
 
-    pub fn set_time(&mut self, time_: Time) {
+    pub fn set_time(&mut self, time_: u32) -> Result<(), BoolError> {
+        let mut c = self.clone();
         unsafe {
-            ffi::g_date_set_time(self.to_glib_none_mut().0, time_);
+            ffi::g_date_set_time_t(c.to_glib_none_mut().0, time_ as _);
         }
-    }
-
-    pub fn set_time_t(&mut self, timet: libc::c_long) {
-        unsafe {
-            ffi::g_date_set_time_t(self.to_glib_none_mut().0, timet);
+        if !Self::valid_dmy(c.get_day(), c.get_month(), c.get_year()) {
+            Err(bool_error!("invalid time"))
+        } else {
+            *self = c;
+            Ok(())
         }
     }
 
     //pub fn set_time_val(&mut self, timeval: /*Ignored*/&mut TimeVal) {
     //    unsafe { TODO: call ffi::g_date_set_time_val() }
     //}
 
-    pub fn set_year(&mut self, year: DateYear) {
-        unsafe {
-            ffi::g_date_set_year(self.to_glib_none_mut().0, year);
+    pub fn set_year(&mut self, year: DateYear) -> Result<(), BoolError> {
+        if !Self::valid_dmy(self.get_day(), self.get_month(), year) {
+            Err(bool_error!("invalid year"))
+        } else {
+            unsafe {
+                ffi::g_date_set_year(self.to_glib_none_mut().0, year);
+            }
+            Ok(())
         }
     }
 
-    pub fn subtract_days(&mut self, n_days: u32) {
-        unsafe {
-            ffi::g_date_subtract_days(self.to_glib_none_mut().0, n_days);
+    pub fn subtract_days(&mut self, n_days: u32) -> Result<(), BoolError> {
+        let julian = self.get_julian();
+        if julian > n_days {
+            Err(bool_error!("invalid number of days"))
+        } else {
+            unsafe {
+                ffi::g_date_subtract_days(self.to_glib_none_mut().0, n_days);
+            }
+            Ok(())
         }
     }
 
-    pub fn subtract_months(&mut self, n_months: u32) {
+    pub fn subtract_months(&mut self, n_months: u32) -> Result<(), BoolError> {
+        // The checks for this function are just a mess in the C code, allowing intermediate
+        // unknonwn state. So for now, nothing can be done...
```

Comment:
```suggestion
        // unknown state. So for now, nothing can be done...
```

`subtract_days()` should be much worse then, but you don't have such a comment there :)

Reply:
Hum no, `subtract_days` is easy to check in comparison: https://gitlab.gnome.org/GNOME/glib/-/blob/master/glib/gdate.c#L1709

Reply:
Ah because it goes through julian days...